contrib/bindings/go/pathrs/pathrs.go

```diff
@@ -0,0 +1,239 @@
+func (r *Root) Hardlink(path, target string) error {
+	status := hardlink(r.root, path, target)
+	if status < 0 {
+		return handleErr(pointerOnRoot, r.root)
+	}
+
+	return nil
+}
+
+// Symlink creates a symlink of file named target and place it to path
+func (r *Root) Symlink(path, target string) error {
+	status := symlink(r.root, path, target)
+	if status < 0 {
+		return handleErr(pointerOnRoot, r.root)
+	}
+
+	return nil
+}
+
+// Free frees underling caught resources
+func (r *Root) Free() {
+	/*
+		Should we handle it in this way or always return not nil pointer on root?
+	*/
+	if r != nil {
+		free(pointerOnRoot, r.root)
+	}
+}
+
+// Handle represents an handle api of rspath
+type Handle struct {
+	handle cRootPtr
+}
+
+// Free frees underling caught resources
+func (h *Handle) Free() {
+	if h == nil {
+		return
+	}
+
+	free(pointerOnHandle, h.handle)
+}
+
+// Reopen upgrade the handle to a file representation
+func (h *Handle) Reopen(flags int) (*os.File, error) {
+	fd := reopen(h.handle, flags)
+	// TODO: verify what is correct path to handle this type of errors
+	// must it always be called error handling as in cat.c file or we can check fd first
+	if err := handleErr(pointerOnHandle, h.handle); err != nil {
+		return nil, err
+	}
+
+	file := os.NewFile(uintptr(fd), "")
+	return file, nil
+}
+
+// Error representation of rust error
+// particularly useful to not frighten to lost controll of pointer which can be rewritten.
+type Error struct {
+	description string
+	backtrace   []backtraceLine
+}
+
+type backtraceLine struct {
+	ip       uintptr
+	sAddress uintptr
+	sName    string
+	sFile    string
+	sLineno  uint32
+}
+
+func (err *Error) Error() string {
+	return err.description
+}
+
+// Backtrace flush backtrace of underlying error to string.
+//
+// Its not passed to realization of Error interface on purpose since
+// the main error should remain clear and simple
+func (err *Error) Backtrace() string {
+	buf := strings.Builder{}
+	buf.Reset()
+
+	for _, line := range err.backtrace {
+		if line.sName != "" {
+			buf.WriteString(fmt.Sprintf("'%s'@", line.sName))
+		}
+		buf.WriteString(fmt.Sprintf("<0x%x>+0x%x\n", line.sAddress, uintptr(line.ip)-uintptr(line.sAddress)))
+		if line.sFile != "" {
+			buf.WriteString(fmt.Sprintf("  in file '%s':%d\n", line.sFile, line.sLineno))
+		}
+	}
+
+	return buf.String()
+}
+
+func newError(err cErrorPtr) *Error {
+	var description string
+	var backtrace []backtraceLine
+
+	realError := (*errorType)(err)
+
+	if realError.savedErrno != 0 {
+		description = fmt.Sprint(syscall.Errno(realError.savedErrno))
+	} else if realError.description != nil {
+		description = toGoString(realError.description)
+	}
+
+	if realError.backtrace != nil {
+		length := uintptr(realError.backtrace.length)
+		size := unsafe.Sizeof(backtraceEntry{})
+		for i := uintptr(0); i < length; i++ {
+			head := (*backtraceEntry)(unsafe.Pointer(uintptr(unsafe.Pointer(realError.backtrace.head)) + size*i))
```

Comment:
This is a bit ugly but I'm not sure how to make it nicer... Oh well, we can fix that later.

Reply:
Absolutely, but I have not seen better solution yet. (The lack of pointer arithmetic cause it or my knowledge :))

Reply:
Here's a nicer way of doing it (see [this playground for an example](https://play.golang.org/p/yswARRHxyVG)):

```go
	head := uintptr(unsafe.Pointer(realError.backtrace.head))
	length := uintptr(realError.backtrace.length)
	sizeof := unsafe.Sizeof(C.backtrace_entry_t{})

	for ptr := head; ptr < head + length*sizeof; ptr += sizeof {
		entry := (*C.backtrace_entry_t)(unsafe.Pointer(ptr))
		// and now put it inside a backtraceEntry
	}
```

The trick is to avoid going to `unsafe.Pointer` until the last possible second, and doing all the arithmetic with `uintptr`. The above loop is effectively the Go equivalent to a similar loop in C.

It's not quite right to cast to a Go struct from a C struct pointer, hence why I've used `C.backtrace_entry_t` above (though I haven't checked if it will compile using those).